src/Lykke.Service.KycReports.Services/Reports/KycReportingService.cs

```diff
@@ -0,0 +1,766 @@
+
+            //var yesterday = DateTime.Today.AddDays(-1);
+            var today = DateTime.Today;
+            if (endDate > today)
+                endDate = today;
+
+            if (endDate < startDate)
+                return new List<string>();
+
+
+            await GenerateKycOfficerStats(startDate, true);
+            var jsonReport = await _reportRepository.GetKycOfficerStatsJsonRows(startDate, endDate);
+
+            return jsonReport.Where(jRow => !jRow.Contains(KycOfficerStatsDataReport.EmptyDayKycOfficer));
+        }
+
+        private async Task<bool> GenerateKycOfficerStats(DateTime from, bool isExcludeExistedRows)
+        {
+            var startDate = from;
+            var endDate = DateTime.Today.AddDays(-1);
+
+            if (startDate > endDate)
+                return false;
+
+            var datesArray = Enumerable.Range(0, 1 + endDate.Subtract(startDate).Days)
+                  .Select(offset => startDate.AddDays(offset))
+                  .ToArray();
+
+            List<KycStatusLogRecord> auditLogEntities = null;
+
+            var isGeneratedOk = true;
+
+            foreach (var startOfDay in datesArray)
+            {
+                try
+                {
+                    if (isExcludeExistedRows)
+                    {
+                        var existedJsonRows = await _reportRepository.GetKycOfficerStatsJsonRows(startOfDay, startOfDay);
+                        if (existedJsonRows.Any())
+                            continue;
+                    }
+
+                    if (auditLogEntities == null)
+                    {
+                        auditLogEntities = await GetKycStatusLogRecords(startDate, endDate);
+                    }
+
+                    var itemsTodayGroups =
+                        auditLogEntities.Where(i => i.Date >= startOfDay && i.Date < startOfDay.AddDays(1))
+                            .OrderBy(i => i.Date)
+                            .GroupBy(i => (KycOfficer: i.KycOfficer, PartnerName: i.PartnerName))
+                            .ToList();
+
+                    if (!itemsTodayGroups.Any())
+                    {
+                        var reportRow = new KycOfficerStatsDataReport()
+                        {
+                            ReportDay = startOfDay,
+                            KycOfficer = KycOfficerStatsDataReport.EmptyDayKycOfficer,
+                            PartnerName = string.Empty,
+                            OnBoardedCount = -1,
+                            DeclinedCount = -1,
+                            ToResubmitCount = -1
+                        };
+
+                        await _reportRepository.InsertRow(reportRow);
+                    }
+
+                    foreach (var itemsToday in itemsTodayGroups)
+                    {
+                        var kycOfficer = itemsToday.Key.KycOfficer;
+                        var partnerName = itemsToday.Key.PartnerName;
+
+                        var onBoardedCount = itemsToday
```

Comment:
Allthese 3 values can be counted within the same foreach loop - to avoid multiple iterations across the same collection

Reply:
rewritten